### ISSUE TITLE
Feature request: Parallel work preserving AD 

### ISSUE THREAD

**Issue by AriMKatz** [JuliaFolds/FLoops.jl#116]
From the [Dex paper](https://arxiv.org/pdf/2104.05372.pdf)

> 5.3.3 Transposing array indexing. The cotangent of array indexing is a recurring problem in the
> design of automatic differentiation systems. The reverse-mode update due to reading x = array.i
> is of course the sparse update array_cot!i += x_cot, but how should we represent this?
> In an imperative language, one is always free to just emit the mutating update, which has been the
> de-facto standard implementation technique of AD systems over decades. This correctly conserves
> work, but turns what used to be a parallelizable loop reading array into a sequential loop writing
> to array_cot. Of course, the parallelism can be recovered by downstream compiler analyses in
> simple cases, but complex cases are not difficult to come by.
> On the other hand, in pure array languages (such as JAX [Bradbury et al. 2018] and other
> Python-based array-libraries) the conventional pattern used to implement reverse-mode AD is
> to just one-hot encode the indexing update, i.e., create an update array of almost all zeros. This
> is simple to implement, and costs no parallelism, but of course creates an asymptotic increase in
> memory usage and count of arithmetic operations. Given the importance of indexing in Dex, this is
> not a viable option for us.
> 
> Another alternative in a pure system is to perform cotangent accumulation functionally by
> accumulating the updates in sparse data structures that gather the index or indices where an update
> is to occur along with the value(s) to be added there. This is also asymptotically work-preserving,
> but imposes large constant or logarithmic factor overhead in performance, as well as a significant
> constant factor in developer effort. It is also quite unfriendly to hardware acceleration due to
> the inherent dynamism in the shape of the data structure and requirements on dynamic memory
> allocation. We are thus not aware of any AD system that actually differentiates indexing this way.
> Instead, we note that while general mutation is certainly sufficient to preserve work, it’s not
> actually necessary—the only mutation that reverse-mode AD needs is to accumulate sums. More
> importantly, because summation is associative (up to floating-point rounding), it can still be executed
> in parallel. The desire to capture this was the prime reason why we introduced an effect system
> into Dex, and in particular why we defined Accum. The Accum effect is expressive enough to capture
> cotangent updates, but restrictive enough to be easy to parallelize (Section 6.1). Having introduced
> Accum into the language, we promptly found other uses for it besides cotangent accumulation, for
> instance the histogram example from Section 6.1.2.

> 5.3.1 Capturing scoped intermediates. Non-linear primitive operations such as mul need to capture
> the intermediate values computed inside the differentiated function. This is fine if the intermediate
> is in scope for the remainder of the function; but whenever we linearize an expression that may
> construct intermediates that go out of scope (such as for), we have to arrange for their values to
> be captured.
> This is why we define differentiation to return core IR rather than the post-simplification subset
> of it. By emitting core IR, linearize and transpose can just capture whatever they need in a fresh
> lambda expression, whose type then doesn’t need to reflect the type of the data it is carrying in
> the closure. We then recover post-simplification IR by running simplification again on the output.
> This architecture does mean that simplification and differentiation have to form a loop in the
> compiler, rather than being sequential passes, but we feel that the simplification of linearization
> and transposition thus won is worth it.
> 

> 5.3.2 Differentiation of state-mutating code. Reverse-mode automatic differentiation is notoriously
> difficult in the presence of mutation in the program being differentiated. The issue is again storage
> of intermediate primal values. But what if one of these primals is mutated by a later operation in
> the program?
> In Dex, this problem disappears, because the type system distinguishes between the (mutable)
> reference being read and/or updated, and the (immutable) value one obtains from it with get.
> 
> The value is saved by linearize as needed, by the same mechanism as all other values. The data
> in the reference itself does not need to be stored, because get is linear: knowing the cotangent of
> the value emitted is enough to compute the update that must be made to the mutable buffer, and
> no additional information about the primal data therein is needed. Conversely, the transpose of
> put just reads the currently accumulated cotangent of the reference.
> Even in pure code, it remains important that only non-linear operations store their inputs. Many
> Dex loops operate on just one or a few index values of the state array per iteration, so making a
> complete copy of the state at each step could raise even the asymptotic complexity of the linearized
> function. Array indexing, however, is linear, so all those copies can be avoided. Indeed, given an
> expression such as for i. sin (get ref).i, only the argument of sin needs to be saved for
> each iteration of the loop.

Here are the chainrules indexing rules: https://github.com/JuliaDiff/ChainRules.jl/blob/3355ae7306f7d624cf9b0c55717cade5ce7865d7/src/rulesets/Base/indexing.jl

**Comment by AriMKatz** [comment]
https://youtu.be/3-uli5fAEps
![Screenshot_20220201-083622](https://user-images.githubusercontent.com/44276225/151978533-a2c3d041-7444-47d8-ba57-428240fb59b1.png)

**Comment by AriMKatz** [comment]
Here's an example of some issues with threadsx https://discourse.julialang.org/t/parallel-reductions-with-zygote/75969/8
